### SySc 575, ECE 4/555, AI: NN1, Fall 2017, Homework #1


Treat this file as a lab notebook, and submit it - including code and written answers - upon completion. 

Give your written responses in ‘markdown’ cells after working out the code. We should be able to execute your code and get the same results that you did. It’s fine to include mistakes and the steps you went through to reach your results. When doing so, clearly mark the “final” code that corresponds to your written answers. Also, the more you annotate and describe your steps and choices, the better you’ll learn and the better we’ll be able to help you. (‘Markdown’ is easy to pick up, if you want to include basic formatting in those cells like I’ve done.)

#############################################################################################
1. Single-layer XOR.
-----------
Adapt the sample code from HW0 to train a single perceptron on XOR data. Examine the resulting prediction, weights, and history, and comment on the success of the training.
#############################################################################################

IMPORTS

In [3]:
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.optimizers import SGD
import numpy as np
import matplotlib.pyplot as plt

Input and Output data definitions

In [4]:
in_data = np.array([[0,0],[0,1],[1,0],[1,1]])
out_data = np.array([[0],[1],[1],[0]])

Single Layer Neural Network Model with sigmoid activation function

In [5]:
model = Sequential()
model.add(Dense(1, input_dim=2, activation='sigmoid'))

In [6]:
model.summary()
print(model.predict_proba(in_data))

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 1)                 3         
Total params: 3
Trainable params: 3
Non-trainable params: 0
_________________________________________________________________
4/4 [==============================] - 0s
[[ 0.5       ]
 [ 0.65112597]
 [ 0.75408173]
 [ 0.85125697]]


In [7]:
error_history = []
weight_history = []

def capture_data(epoch,logs):
    weights = model.get_weights()
    weight_history.append(weights)
    error_history.append(logs['loss'])
    
capture_data_callback = keras.callbacks.LambdaCallback(on_epoch_end=capture_data)

Training the Neural Network model

In [15]:
sgd = SGD(lr=0.8)
model.compile(loss='mse', optimizer=sgd)
hist = model.fit(in_data, out_data, batch_size=1, epochs=10000, verbose=0,callbacks=[capture_data_callback])

Running test data

In [16]:
print(model.predict_proba(in_data))
print(model.get_weights())

4/4 [==============================] - 0s
[[ 0.49528703]
 [ 0.462486  ]
 [ 0.45628721]
 [ 0.42389971]]
[array([[-0.15644632],
       [-0.1314861 ]], dtype=float32), array([-0.01885244], dtype=float32)]


Since the XOR data is not linearly separable in a two-dimensional weight space, a single neuron with 2-inputs cannot be reliably used for a XOR logic implementation.
Varying the learning rate, number of epochs don't improve the accuracy of the trained model. 

#############################################################################################
2. XOR with a hidden layer of 2 nodes.
-------------
Repeat the training process with a multilayer perceptron, using two nodes in a hidden layer.
* Identify the output weights, and describe what is happening with the hidden nodes and the output node.
* Do the fit process at least a few more times, starting with new models and random seeds. What is different or the same about each of these experiments?
* For two of the runs that appear substantially different in their weights, identify the logical function that each intermediary node is computing, as well as the output node. (For a network like this, with two inputs and one output per node, you should be able to write out a logic sentence or fill in a table of binary values.)

#############################################################################################

Input and Output data definitions

In [54]:
in_data = np.array([[0,0],[0,1],[1,0],[1,1]])
out_data = np.array([[0],[1],[1],[0]])

Single Layer Neural Network Model with sigmoid activation function

In [55]:
model = Sequential()
model.add(Dense(2, input_dim=2, activation='sigmoid',kernel_initializer=keras.initializers.glorot_uniform(seed=348)))
model.add(Dense(1, activation='sigmoid',kernel_initializer=keras.initializers.glorot_uniform(seed=847)))

In [56]:
model.summary()
print(model.predict_proba(in_data))
print(model.get_weights())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_10 (Dense)             (None, 2)                 6         
_________________________________________________________________
dense_11 (Dense)             (None, 1)                 3         
Total params: 9
Trainable params: 9
Non-trainable params: 0
_________________________________________________________________
4/4 [==============================] - 0s
[[ 0.61063296]
 [ 0.61387116]
 [ 0.60049093]
 [ 0.60314268]]
[array([[ 0.41417965, -0.61161131],
       [-0.92264259,  0.71484089]], dtype=float32), array([ 0.,  0.], dtype=float32), array([[ 0.36344996],
       [ 0.53649718]], dtype=float32), array([ 0.], dtype=float32)]


In [57]:
error_history = []
weight_history = []

def capture_data(epoch,logs):
    weights = model.get_weights()
    weight_history.append(weights)
    error_history.append(logs['loss'])
    
capture_data_callback = keras.callbacks.LambdaCallback(on_epoch_end=capture_data)

Training the Neural Network model

In [62]:
sgd = SGD(lr=0.9, momentum=0.8)
model.compile(loss='mse', optimizer=sgd)
hist = model.fit(in_data, out_data, batch_size=1, epochs=10000, verbose=0,callbacks=[capture_data_callback])

Run test data

In [63]:
print("Prediction Probability:")
print(model.predict_proba(in_data))
print ("Weights:")
print(model.get_weights())

Prediction Probability:
4/4 [==============================] - 0s
[[ 0.00248843]
 [ 0.99785662]
 [ 0.99785644]
 [ 0.00225813]]
Weights:
[array([[ 7.05005074, -7.27228308],
       [-7.25423193,  7.07936239]], dtype=float32), array([-3.84114361, -3.85897756], dtype=float32), array([[ 13.19900513],
       [ 13.1932354 ]], dtype=float32), array([-6.54351759], dtype=float32)]


**1. Hidden Layer Seed: None, Output Layer Seed: None, Learning Rate: 0.6**<br>
Prediction Probability:
[[ 0.01052439]
 [ 0.99011272]
 [ 0.98792142]
 [ 0.00943174]]<br>
Weights:
[array([[-6.05538273, -6.64251518],
       [ 6.28175402,  6.57256651]], dtype=float32), array([ 3.02769852, -3.55320954], dtype=float32), array([[-10.17666912],
       [ 10.3923254 ]], dtype=float32), array([ 4.87385082], dtype=float32)]

**2. Hidden Layer Seed: 15, Output Layer Seed: 87,Learning Rate: 0.6**<br>
Prediction Probability:
[[ 0.00963103]
 [ 0.98992419]
 [ 0.98997027]
 [ 0.01235967]]<br>
Weights:
[array([[-6.74988985, -5.14753294],
       [-6.67412758, -5.13247728]], dtype=float32), array([ 2.79323983,  7.66390371], dtype=float32), array([[-10.84962273],
       [ 10.70538902]], dtype=float32), array([-5.10975122], dtype=float32)]

**3. Hidden Layer Seed: 348, Output Layer Seed: 847, Learning Rate: 0.6**<br>
Prediction Probability:
[[ 0.01146676]
 [ 0.99012995]
 [ 0.99012685]
 [ 0.0102966 ]]<br>
Weights:
[array([[ 6.18384171, -6.38674974],
       [-6.37823915,  6.2049284 ]], dtype=float32), array([-3.41661978, -3.43107319], dtype=float32), array([[ 10.32640266],
       [ 10.32269287]], dtype=float32), array([-5.10843992], dtype=float32)]

**4. Hidden Layer Seed: 348, Output Layer Seed: 847, Learning Rate: 0.6, Momentum: 0.9**<br>
Prediction Probability:
[[ 0.00329878]
 [ 0.99715865]
 [ 0.99715841]
 [ 0.00298746]]<br>
Weights:
[array([[ 6.89499903, -7.10108995],
       [-7.09406614,  6.90929937]], dtype=float32), array([-3.76589823, -3.77518201], dtype=float32), array([[ 12.67327595],
       [ 12.67069435]], dtype=float32), array([-6.28166866], dtype=float32)]

For Experiment 2, Logic function table:<br>
Neuron 1<br>
0 0 : 1<br>
0 1 : 0<br>
1 0 : 0<br>
1 1 : 0<br>
Neuron 2<br>
0 0 : 1<br>
0 1 : 1<br>
1 0 : 1<br>
1 1 : 0<br>
Neuron 3<br>
0 0 : 0<br>
0 1 : 1<br>
1 0 : 1<br>
1 1 : 0<br>

For Experiment 4, Logic function table:<br>
Neuron 1<br>
0 0 : 1<br>
0 1 : 0<br>
1 0 : 0<br>
1 1 :<br>
Neuron 2<br>
0 0 : 1<br>
0 1 : 1<br>
1 0 : 1<br>
1 1 : 0<br>
Neuron 3<br>
0 0 : 0<br>
0 1 : 1<br>
1 0 : 1<br>
1 1 : 0<br>


#############################################################################################
3. Linearly separable logic with hidden layer
---------------
Train a network with the same structure as in Problem 2 on an easier logic relation of your choice – one that is linearly separable, and for which the hidden nodes are not necessary.
* Before fitting the model, write down your predictions or expectations of how the training process might be different, and how the network will perform.
* After training, see if you are able to interpret the weights, and what the hidden nodes are computing.
* How did this experiment go relative to your predictions? Was there anything different about the training process?

#############################################################################################

Input and Output data definitions

In [64]:
in_data = np.array([[0,0],[0,1],[1,0],[1,1]])
out_data = np.array([[0],[0],[1],[1]])

Single Layer Neural Network Model with sigmoid activation function

In [77]:
model = Sequential()
model.add(Dense(2, input_dim=2, activation='sigmoid',kernel_initializer=keras.initializers.glorot_uniform(seed=874)))
model.add(Dense(1, activation='sigmoid',kernel_initializer=keras.initializers.glorot_uniform(seed=247)))

model = Sequential()
model.add(Dense(2, input_dim=2, activation='sigmoid'))
model.add(Dense(1, activation='sigmoid'))

In [78]:
model.summary()
print(model.predict_proba(in_data))
print(model.get_weights())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_16 (Dense)             (None, 2)                 6         
_________________________________________________________________
dense_17 (Dense)             (None, 1)                 3         
Total params: 9
Trainable params: 9
Non-trainable params: 0
_________________________________________________________________
4/4 [==============================] - 0s
[[ 0.62251264]
 [ 0.66056901]
 [ 0.63159949]
 [ 0.66820812]]
[array([[ 0.44191486,  0.17432195],
       [-0.67715901,  0.59169042]], dtype=float32), array([ 0.,  0.], dtype=float32), array([[-0.07099704],
       [ 1.07145059]], dtype=float32), array([ 0.], dtype=float32)]


In [79]:
error_history = []
weight_history = []

def capture_data(epoch,logs):
    weights = model.get_weights()
    weight_history.append(weights)
    error_history.append(logs['loss'])
    
capture_data_callback = keras.callbacks.LambdaCallback(on_epoch_end=capture_data)

Training the Neural Network model

In [80]:
sgd = SGD(lr=0.5)
model.compile(loss='mse', optimizer=sgd)
hist = model.fit(in_data, out_data, batch_size=1, epochs=10000, verbose=0,callbacks=[capture_data_callback])

Run test data

In [83]:
print(model.predict_proba(in_data))
optimized_weights = model.get_weights()
print(optimized_weights)

4/4 [==============================] - 0s
[[ 0.00948516]
 [ 0.00821553]
 [ 0.99332166]
 [ 0.99225444]]
[array([[ 1.25975549,  6.69919395],
       [-0.68281335, -0.18417738]], dtype=float32), array([-0.39594024, -3.24765706], dtype=float32), array([[  0.55842936],
       [ 10.17581749]], dtype=float32), array([-5.2538352], dtype=float32)]


The logic function that I have chosen for follows Input 1. It is lineraly separable and should therefore the model should perform quite well. Input 1 should see higher weights that Input 2.<br>

**1. Hidden Layer Seed: 24, Output Layer Seed: 97**<br>
Weights:<br>
[array([[-5.66260767,  3.43532205],
       [ 0.24899971, -0.06949561]], dtype=float32), array([ 2.40411735, -1.28243279], dtype=float32), array([[-8.20691872],
       [ 4.54485273]], dtype=float32), array([ 1.43486977], dtype=float32)]<br>
**2. Hidden Layer Seed: 874, Output Layer Seed: 247**<br>
Weights:<br>
[array([[ 1.25975549,  6.69919395],
       [-0.68281335, -0.18417738]], dtype=float32), array([-0.39594024, -3.24765706], dtype=float32), array([[  0.55842936],
       [ 10.17581749]], dtype=float32), array([-5.2538352], dtype=float32)]<br>
       
As we can see from the above set of final weights, Input 1 does not have a higher weight that Input 2.<br>
Patterns I see from the weights: In hidden layer, Neuron 1 has higher BIAS weight thatn Neuron 2<br>



#############################################################################################
4. XOR in a larger network
----------
Create a new network structure to train on the XOR data. For instance, try using additional hidden nodes or another hidden layer. For any change you make, predict what the effect will be on the experiment relative to the outcomes in Problem 2. Analyze the results and compare them to your predictions.
#############################################################################################

Input and Output data definitions

In [91]:
in_data = np.array([[0,0],[0,1],[1,0],[1,1]])
out_data = np.array([[0],[1],[1],[0]])

Single Layer Neural Network Model with sigmoid activation function

In [129]:
model = Sequential()
model.add(Dense(4, input_dim=2, activation='sigmoid'))
model.add(Dense(5, activation='sigmoid'))
model.add(Dense(6, activation='sigmoid'))
model.add(Dense(1, activation='sigmoid'))

model = Sequential()
model.add(Dense(2, input_dim=2, activation='sigmoid'))
model.add(Dense(1, activation='sigmoid'))

In [130]:
model.summary()
print(model.predict_proba(in_data))
print(model.get_weights())

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_38 (Dense)             (None, 4)                 12        
_________________________________________________________________
dense_39 (Dense)             (None, 5)                 25        
_________________________________________________________________
dense_40 (Dense)             (None, 6)                 36        
_________________________________________________________________
dense_41 (Dense)             (None, 1)                 7         
Total params: 80
Trainable params: 80
Non-trainable params: 0
_________________________________________________________________
4/4 [==============================] - 0s
[[ 0.6920554 ]
 [ 0.69246233]
 [ 0.69203466]
 [ 0.69245374]]
[array([[ 0.40737545, -0.5712862 ,  0.87423062,  0.08292985],
       [ 0.25969815,  0.7765348 ,  0.31058502, -0.03965157]], dtype=float32), array([ 0.,  0.,  0.,  0.], dtype=f

In [131]:
error_history = []
weight_history = []

def capture_data(epoch,logs):
    weights = model.get_weights()
    weight_history.append(weights)
    error_history.append(logs['loss'])
    
capture_data_callback = keras.callbacks.LambdaCallback(on_epoch_end=capture_data)

Training the Neural Network model

In [132]:
sgd = SGD(lr=0.5)
model.compile(loss='mse', optimizer=sgd)
hist = model.fit(in_data, out_data, batch_size=1, epochs=10000, verbose=0,callbacks=[capture_data_callback])

Run test data

In [133]:
print(model.predict_proba(in_data))
optimized_weights = model.get_weights()
print(optimized_weights)

4/4 [==============================] - 0s
[[ 0.48947161]
 [ 0.48980319]
 [ 0.48986641]
 [ 0.49010268]]
[array([[ 0.6474157 , -0.58169234,  0.91257459,  0.22898449],
       [ 0.53961569,  0.74498135,  0.4035387 ,  0.1236072 ]], dtype=float32), array([ 0.2045404 , -0.06186254,  0.00787427,  0.18724956], dtype=float32), array([[ 0.73662692,  0.46659914,  0.11046245,  0.65352261,  0.48172688],
       [ 0.17069897,  0.11714319, -0.56289881, -0.28468245,  0.70778698],
       [-0.0616717 ,  0.88323247,  0.32364142,  0.0049238 , -0.15383676],
       [ 0.84545815, -0.16861406, -0.66328788,  1.07056141,  0.30567119]], dtype=float32), array([ 0.38340509,  0.63455564,  0.00930588,  0.74555218,  0.34990939], dtype=float32), array([[-0.21805637, -1.32631421, -0.39851546, -1.10291934, -0.04725088,
        -1.0902828 ],
       [-0.85824841, -0.95033211, -0.83005536, -0.95987159, -1.31712162,
        -0.52558774],
       [-0.05303051, -0.84969211,  0.32949066, -0.49902901,  0.04891429,
        -0.40390

increasing the number of nodes in the hidden layer should improve accuracy.

**Hidden Layer 1: 2 nodes**<br>
[[ 0.01064747]
 [ 0.98884654]
 [ 0.98888898]
 [ 0.01369623]]<br>
**Hidden Layer 1: 5 nodes**<br>
[[ 0.00679121]
 [ 0.99107158]
 [ 0.99153799]
 [ 0.01047783]]<br>
**Hidden Layer 1: 4 nodes**<br>
[[ 0.0069016 ]
 [ 0.99095839]
 [ 0.99178982]
 [ 0.01030412]]<br>
**Hidden Layer 1: 2 nodes, Hidden Layer 2: 4 n3des**<br>
[[ 0.01026452]
 [ 0.99129486]
 [ 0.98990965]
 [ 0.00926109]]<br>
**Hidden Layer 1: 5 nodes, Hidden Layer 2: 4 nodes, Hidden Layer 1: 1 nodes**<br>
[[ 0.00999548]
 [ 0.98751265]
 [ 0.99064857]
 [ 0.00872787]]
**Hidden Layer 1: 4 nodes, Hidden Layer 2: 5 nodes, Hidden Layer 1: 1 nodes**<br>
[[ 0.00651652]
 [ 0.99047595]
 [ 0.99042261]
 [ 0.00914785]]


As we see from the data above,increase in number of nodes per hidden layer increases the accuracy. But adding more hidden layers don't seem to have the same effect.

